### PR TITLE
Add cache to Eth1 RPC functions

### LINKED ISSUE #5340: BlockFetcher RPC functions are inefficient
After observation from Nishant and I on the RPC functions used when the beacon chain has to make any block queries. We've noticed that there are a few problems possibly related to #5268.

[BlockNumberByTimestamp](https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/powchain/block_reader.go#L75) is extremely inefficient, it goes down from the head of the chain and backwards block-by-block until it finds a block timestamp that fits the requested timestamp.

[BlockTimeByHeight](https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/powchain/block_reader.go#L61) does not use the cache for blocks at all.

After these improvements are complete, we should see a considerable speedup in interactions with ETH1.

### PR DESCRIPTION
**Pull request by 0xKiwi**
Resolves #5340 

This PR improves the RPC functions by adding a cache to both.

### CODE REVIEW

File: beacon-chain/powchain/block_reader.go [hunk 1]
```diff
@@ -72,8 +80,9 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
 // This is a naive implementation that will use O(ETH1_FOLLOW_DISTANCE) calls to cache
 // or ETH1. This is called for multiple times but only changes every
 // SlotsPerEth1VotingPeriod (1024 slots) so the whole method should be cached.
+// How should caching the whole method be done?
```
Comment (nisdas) [inline beacon-chain/powchain/block_reader.go:83]: ? do we need this comment

Comment (0xKiwi) [reply beacon-chain/powchain/block_reader.go:83]: Just for personal tracking as I try to figure out whats needed, will remove before the PR is ready for merging

File: beacon-chain/rpc/validator/proposer.go [hunk 2]
```diff
@@ -29,6 +29,8 @@ import (
 
 // eth1DataNotification is a latch to stop flooding logs with the same warning.
 var eth1DataNotification bool
+var cachedEth1VotingStartTime uint64
+var cachedEth1Data *ethpb.Eth1Data
```
Comment (0xKiwi) [inline beacon-chain/rpc/validator/proposer.go:33]: This caching is very simple, should we cache it on the beacon node side or is on the validator end fine?

Comment (rauljordan) [reply beacon-chain/rpc/validator/proposer.go:33]: It should be fine if it lives in the beacon node side, that way many validator clients connected to one beacon node can benefit

**Pull request merged**